snooty/postprocess.py

```diff
@@ -919,6 +919,57 @@ def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
             self.guides[current_slug].description = node.children
 
 
+class OpenAPIHandler(Handler):
+    """Constructs metadata for OpenAPI content pages."""
+
+    @dataclass
+    class SourceData:
+        source_type: str
+        source: str
+
+    def __init__(self, context: Context) -> None:
+        super().__init__(context)
+        self.openapi_pages: Dict[str, OpenAPIHandler.SourceData] = {}
+
+    def get_metadata(self) -> Dict[str, SerializableType]:
+        """Returns serialized object to be used as part of the build's metadata."""
+
+        return {k: asdict(v) for k, v in self.openapi_pages.items()}
+
+    def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        if (
+            not isinstance(node, n.Directive)
+            or node.name != "openapi"
+            or node.options.get("preview")
+        ):
+            return
+
+        current_file = fileid_stack.current
+        current_slug = clean_slug(current_file.without_known_suffix)
+
+        if current_slug in self.openapi_pages:
+            self.context.diagnostics[current_file].append(
+                DuplicateDirective(node.name, node.start[0])
+            )
+            return
+
+        # source_type should be assigned in the parsing layer
+        source_type = node.options.get("source_type")
+        if not source_type:
+            return
+
+        source = ""
+        argument = node.argument[0]
+        if source_type == "local" or source_type == "atlas":
+            assert isinstance(argument, n.Text)
```

Comment:
is this guaranteed to be true? i.e. can a writer subvert this?

Reply:
This should be guaranteed to be true (unless there's some edge case I'm missing). The parser handles setting the `source_type` ([example](https://github.com/mongodb/snooty-parser/pull/426/files#diff-5e66e8539da7ed36bf30c7a27807d04eb3c8c9a0d78649afcdc4a90cb1871efeR715)), which is dependent on the argument and its node structure anyways. This logic in the postprocessor echoes that to grab the correct source without needing to go through similar parser logic to re-determine the argument's structure.

I also don't think a writer can subvert this. Since `source_type` isn't an exposed option in `rstspec`, the parser throws an error when attempting to mix and match `source_type` and the argument.

Reply:
Yep makes sense! Thank you!

Could you put that in a comment? After that, lgtm!

Reply:
Done! https://github.com/mongodb/snooty-parser/pull/426/files#diff-d321f31a0d6e1148a3d4e4ec2fded3b660d822fcd629909c38fcc30dd3bc493eR963-R967